tests/OTAnalytics/application/test_state.py

```diff
@@ -145,3 +153,63 @@ def test_update_image(self) -> None:
         assert track_view_state.background_image.get() == background_image
 
         plotter.plot.assert_called_once()
+
+
+class TestTracksMetadata:
+    @pytest.fixture
+    def first_detection(self) -> Mock:
+        first_detection = Mock(spec=Detection).return_value
+        first_detection.occurrence = datetime(2000, 1, 1, 7, 0, 0, 0)
+        return first_detection
+
+    @pytest.fixture
+    def second_detection(self) -> Mock:
+        second_detection = Mock(spec=Detection).return_value
+        second_detection.occurrence = datetime(2000, 2, 1, 0, 0, 0, 0)
+        return second_detection
+
+    @pytest.fixture
+    def third_detection(self) -> Mock:
+        third_detection = Mock(spec=Detection).return_value
+        third_detection.occurrence = datetime(2000, 2, 5, 0, 0, 0, 0)
+        return third_detection
+
+    @patch("OTAnalytics.application.state.TracksMetadata._get_all_track_detections")
+    def test_update_detection_occurrences(
+        self,
+        mock_get_all_track_detections: Mock,
+        first_detection: Mock,
+        second_detection: Mock,
+        third_detection: Mock,
+    ) -> None:
+        mock_track_repository = Mock(spec=TrackRepository)
+
+        mock_get_all_track_detections.return_value = [
+            first_detection,
+            third_detection,
+            second_detection,
+        ]
+        tracks_metadata = TracksMetadata(mock_track_repository)
+
+        assert tracks_metadata.first_detection_occurrence is None
+        assert tracks_metadata.last_detection_occurrence is None
+
+        tracks_metadata.notify_tracks([])
+        assert tracks_metadata.first_detection_occurrence == first_detection.occurrence
+        assert tracks_metadata.last_detection_occurrence == third_detection.occurrence
+
+        mock_get_all_track_detections.assert_called_once()
+
+    def test_get_all_track_detections(
+        self, first_detection: Mock, second_detection: Mock
+    ) -> None:
+        track = Mock(spec=Track).return_value
```

Comment:
Should the `return_value` part be added to `track.detections` instead?

Reply:
When trying to modify the mock objects members as you suggested will raise an error.
It's kind of hard to explain. Mocking methods will work as expected. But when it comes to mocking the objects members it is a little bit different. In this case the `Mock(spec=Track).return_value` is needed and  will return a mock instance that you can edit.

See: https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.return_value